guardian/pkg/asyncutil/async.go

```diff
@@ -0,0 +1,186 @@
+
+type CommandDispatcher []interface {
+	PauseExecution()
+	ResumeExecution()
+	ShutdownSignaler() Signaler
+}
+
+func (dispatcher CommandDispatcher) PauseExecution() {
+	for _, executor := range dispatcher {
+		executor.PauseExecution()
+	}
+}
+
+func (dispatcher CommandDispatcher) ResumeExecution() {
+	for _, executor := range dispatcher {
+		executor.ResumeExecution()
+	}
+}
+
+func (dispatcher CommandDispatcher) WaitForShutdown() {
+	for _, executor := range dispatcher {
+		<-executor.ShutdownSignaler().Receive()
+	}
+}
+
+type commandExecutor[Req any, Resp any] struct {
+	command             func(context.Context, Req) (Resp, error)
+	pauseExecution      chan chan struct{}
+	resumeExecution     Signaler
+	cmdChan             chan Command[Req, Resp]
+	backlogChan         chan Command[Req, Resp]
+	inflightCmds        sync.WaitGroup
+	executeSig          Signaler
+	shutdownCompleteSig Signaler
+	errBuff             AsyncErrorBuffer
+}
+
+// NewCommandExecutor creates a new RequestHandler implementation.
+func NewCommandExecutor[C any, R any](ctx context.Context, errBuff AsyncErrorBuffer, f func(context.Context, C) (R, error)) CommandExecutor[C, R] {
+	hdlr := &commandExecutor[C, R]{
+		command:         f,
+		errBuff:         errBuff,
+		cmdChan:         make(chan Command[C, R], 100),
+		backlogChan:     make(chan Command[C, R], 100),
+		pauseExecution:  make(chan chan struct{}),
+		resumeExecution: NewSignaler(),
+
+		executeSig:          NewSignaler(),
+		shutdownCompleteSig: NewSignaler(),
+	}
+
+	go hdlr.loop(ctx)
+	return hdlr
+}
+
+func (executor *commandExecutor[C, R]) Receive() <-chan Command[C, R] {
+	return executor.cmdChan
+}
+
+func (executor *commandExecutor[C, R]) ShutdownSignaler() Signaler {
+	return executor.shutdownCompleteSig
+}
+
+func (executor *commandExecutor[C, R]) Send(params C) <-chan Result[R] {
+	cmd, resultChan := NewCommand[C, R](params)
+	executor.cmdChan <- cmd
+	return resultChan
+}
+
+func (executor *commandExecutor[C, R]) loop(shutdownCtx context.Context) {
+	var backlog []Command[C, R]
+
+	defer close(executor.cmdChan)
+	defer close(executor.backlogChan)
+	defer executor.executeSig.Close()
+	defer close(executor.pauseExecution)
+	defer executor.resumeExecution.Close()
+	ctx, stopCommands := context.WithCancel(shutdownCtx)
+	defer func() {
+		defer stopCommands()
+		defer executor.shutdownCompleteSig.Close()
+
+		logrus.Debug("Returning errors for outstanding requests due to shutdown.")
+		for _, req := range backlog {
+			req.ReturnError(context.Canceled)
+		}
+		logrus.Debug("Finished returning errors for outstanding requests due to shutdown.")
+	}()
+
+	var pause bool
+	for {
+		select {
+		case <-shutdownCtx.Done():
+			logrus.Debug("Shutdown signal received, shutting down...")
+			return
+		case cmd := <-executor.cmdChan:
+			logrus.Debugf("Received command.")
+			if !pause {
+				executor.executeCommand(ctx, cmd)
+			} else {
+				backlog = append(backlog, cmd)
+			}
+		case cmd := <-executor.backlogChan:
```

Comment:
Can we have more than 100 tasks in the backlog channel ?

Reply:
Not at the same time. Since this loop never blocks it's unlikely we'd every reach that level of requests, and if we do it would be cleared out.